### PR TITLE
fix(password): fix password confirmation reset and validation failures

### PR DESCRIPTION
**Pull request by cerebrl**
- ensure password confirmation gets reset upon submission failure
- ensure confirmation message is correct after server validation failure
- improve messaging when policy failures has 0 for character sets

Confirmation improvements can be tested here: https://preview-pr-94-ryanbas21.cloud.okteto.net/e2e/widget/inline?journey=TEST_LoginWithStageAttributes

### CODE REVIEW

**Comment by ryanbas21** [comment]
Your preview environment [pr-94-ryanbas21](https://cloud.okteto.com/#/previews/pr-94-ryanbas21) has been deployed.

  Preview environment endpoint is available [here](https://preview-pr-94-ryanbas21.cloud.okteto.net)

File: src/lib/components/_utilities/locale-strings.story.svelte [hunk 1]
```diff
@@ -1,8 +1,11 @@
...
   import Centered from '$components/primitives/box/centered.svelte';
+  import Text from '$components/primitives/text/text.svelte';
   import T from '$components/_utilities/locale-strings.svelte';
 </script>
 
 <Centered>
-  <T key="continueWith" />
+  <Text>
+    <T key="continueWith" />
+  </Text>
```
Comment (cerebrl) [inline src/lib/components/_utilities/locale-strings.story.svelte:10]: This story and the one below just needed to be wrapped so it had the proper theme treatment in the rendered storybook component.

File: src/lib/journey/callbacks/_utilities/callback.utilities.ts [hunk 2]
```diff
@@ -69,16 +69,50 @@ export function getInputTypeFromPolicies(policies: StringDict<unknown>): 'email'
 export function getValidationFailureParams(
   failedPolicy: PolicyRequirement | undefined,
 ): RestructuredParam[] {
-  if (failedPolicy?.policyRequirement === 'CHARACTER_SET') {
-    const params = failedPolicy?.params as {
+  if (failedPolicy?.policyRequirement === 'DICTIONARY') {
```
Comment (cerebrl) [inline src/lib/journey/callbacks/_utilities/callback.utilities.ts:72]: This adds a few policies that were missing to the `ValidatedCreatePasswordCallback` that Stoyan found.

File: src/lib/journey/callbacks/_utilities/callback.utilities.ts [hunk 3]
```diff
@@ -69,16 +69,50 @@ export function getInputTypeFromPolicies(policies: StringDict<unknown>): 'email'
...
+        message: interpolate('passwordCannotContainCommonPasswordsOrBeReversible'),
+        rule: 'reversibleSubstrings',
+      });
+    } else {
+      arr.push({
+        length: null,
+        message: interpolate('passwordCannotContainCommonPasswords'),
+        rule: 'reversibleSubstrings',
+      });
+    }
```
Comment (cerebrl) [inline src/lib/journey/callbacks/_utilities/callback.utilities.ts:103]: I hope I translated the actual rules into human understandable text.

File: src/lib/journey/callbacks/_utilities/callback.utilities.ts [hunk 4]
```diff
@@ -271,32 +329,64 @@ function convertCharacterSetToRuleObj(set: string) {
   const type = arr[1];
 
   if (type === '0123456789') {
-    return {
-      length: Number(num),
-      message: interpolate('minimumNumberOfNumbers', { num: String(num) }),
-      rule: 'numbers',
-    };
+    if (num === '0') {
```
Comment (cerebrl) [inline src/lib/journey/callbacks/_utilities/callback.utilities.ts:332]: All these new checks handle the rule when the num is `0`. Of course, we can't communicate to the user that this requires a character set, but say at least `0` are needed. So, this handles this case with a bit more elegance.

File: src/lib/journey/callbacks/password/confirm-input.svelte [hunk 5]
```diff
@@ -24,14 +25,26 @@
 
   let isVisible = false;
   let type: 'password' | 'text' = 'password';
-  let value: unknown;
+  let value: Maybe<string>;
+
+  function onChangeWrapper(event: Event) {
```
Comment (cerebrl) [inline src/lib/journey/callbacks/password/confirm-input.svelte:30]: Creating a wrapper here to grab the value of the input before passing it to the parent `onChange` function.

File: src/lib/journey/callbacks/password/confirm-input.svelte [hunk 6]
```diff
@@ -24,14 +25,26 @@
...
   /**
    * @function toggleVisibility - toggles the password from masked to plaintext
    */
   function toggleVisibility() {
     isVisible = !isVisible;
     type = isVisible ? 'text' : 'password';
   }
+
+  $: {
+    if (resetValue) {
```
Comment (cerebrl) [inline src/lib/journey/callbacks/password/confirm-input.svelte:43]: There's a bit of logic in the parent component that helps inform this component of when to clear out its value.

File: src/lib/journey/callbacks/password/validated-create-password.mock.ts [hunk 7]
```diff
@@ -157,6 +157,43 @@ export default {
...
+            ],
+            conditionalPolicies: null,
+          },
+        },
+        {
+          name: 'failedPolicies',
+          value: [
+            // TODO: Fix validation crash
+            '{ "policyRequirement": "LENGTH_BASED", "params": { "max-password-length": 0, "min-password-length": 8 } }',
+            '{ "policyRequirement": "CHARACTER_SET", "params": { "allow-unclassified-characters": true, "character-set-ranges": [  ], "character-sets": [ "0:0123456789", "0:ABCDEFGHIJKLMNOPQRSTUVWXYZ", "0:abcdefghijklmnopqrstuvwxyz", "0:~!@#$%^&*()-_=+[]{}|;:,.<>/?" ], "min-character-sets": 4 } }',
```
Comment (cerebrl) [inline src/lib/journey/callbacks/password/validated-create-password.mock.ts:185]: Here's the new `0` number for the character sets. Notice the flag `min-character-sets`.

File: src/lib/journey/callbacks/password/validated-create-password.stories.js [hunk 8]
```diff
@@ -45,6 +45,12 @@ export const PolicyErrors = {
   },
 };
 
+export const PolicyErrorsAlt = {
```
Comment (cerebrl) [inline src/lib/journey/callbacks/password/validated-create-password.stories.js:48]: Adding a story to handle this new `0` case.

File: src/locales/ca/en/index.json
Comment (cerebrl) [inline src/locales/ca/en/index.json:1]: Let me know if these messages make sense. CC @forgerock-chris

File: tests/widget/inline/widget-inline.stage-attribute.test.js [hunk 9]
```diff
@@ -9,13 +9,34 @@ test('Inline widget with additional stage attribute testing', async ({ page }) =
...
+  // Ensure error message is removed
+  await expect(page.getByText('Passwords do not match')).toBeHidden();
+
+  await expect(page.getByLabel('Password', { exact: true })).toHaveValue('');
+  await expect(page.getByLabel('Confirm password')).toHaveValue('');
+
+  await page.getByLabel('Password', { exact: true }).fill('j56eKtae*1');
+  await page.getByLabel('Confirm password').fill('not a match');
+  await page.keyboard.press('Tab'); // focuses off confirm password
+
```
Comment (cerebrl) [inline tests/widget/inline/widget-inline.stage-attribute.test.js:39]: Adding more testing to the confirm password around client and server validation failures, ensuring we correctly report even after successful client validation, but failed server validation.

File: src/lib/journey/callbacks/_utilities/callback.utilities.ts [hunk 10]
```diff
@@ -69,16 +69,50 @@ export function getInputTypeFromPolicies(policies: StringDict<unknown>): 'email'
 export function getValidationFailureParams(
   failedPolicy: PolicyRequirement | undefined,
 ): RestructuredParam[] {
-  if (failedPolicy?.policyRequirement === 'CHARACTER_SET') {
-    const params = failedPolicy?.params as {
+  if (failedPolicy?.policyRequirement === 'DICTIONARY') {
+    const params = failedPolicy.params as {
```
Comment (ryanbas21) [inline src/lib/journey/callbacks/_utilities/callback.utilities.ts:73]:
```suggestion
    const params = failedPolicy.params satisfies {
```

Comment (ryanbas21) [reply src/lib/journey/callbacks/_utilities/callback.utilities.ts:73]: nvm doesnt work here

**Review by ryanbas21 — APPROVED** [review]
overall is fine, im not hard sold on passing an undefined to an onChange handler, but its also just fine.

**Pull request merged**